### PR TITLE
RIT-191

### PR DESCRIPTION
**Pull request by jrocca82**
Fix vesting contract and global issues. Vault contract coming on a different PR.
Ignored issues A3, B3, and B4.
RenounceOwnership issue on RIT-220 PR

Unit tests for contract balance are on RIT-220 PR

Audit Report:
https://docs.google.com/document/d/14rzwHz1Z8Wtir7xRqILzXX4YDcK0QJ_EeWWIxr-eOf8/edit?usp=sharing

### CODE REVIEW

**Review by amagon270 — APPROVED** [review]

File: contracts/SaleVesting.sol [hunk 1]
```diff
@@ -112,8 +122,15 @@ contract SaleVesting is Ownable {
                 _vestingDetails[i].claimStartTime
             );
 
+            totalVestingAmount +=
```
Comment (xiangzhe0812) [inline contracts/SaleVesting.sol:125]: what's this for?
if you want totalVestingAmount, it should be 
totalVestingAmount + =_vestingDetails[i].vestingAmount;

File: contracts/TeamVesting.sol [hunk 2]
```diff
@@ -31,15 +30,22 @@ struct VestingDetail {
...
+        deployDate = block.timestamp;
         //Vesting start date
         startDate = _startDate;
+        require(
+            startDate >= deployDate,
+            "Start date cannot be before the deployment date"
+        );
+        self = address(this);
+        RITE = _RITE;
+        require(_RITE != address(0), "Address cannot be zero");
```
Comment (xiangzhe0812) [inline contracts/TeamVesting.sol:48]:
```suggestion
       require(
            _startDate >= block.timestamp,
            "Start date cannot be before the deployment date"
        );
        require(_RITE != address(0), "Address cannot be zero");
        self = address(this);
        RITE = _RITE;
  
```

File: contracts/TeamVesting.sol [hunk 3]
```diff
@@ -31,15 +30,22 @@ struct VestingDetail {
 contract TeamVesting is Ownable {
```
Comment (xiangzhe0812) [inline contracts/TeamVesting.sol:30]: 1. Check there are tokens available
2. Need add totalClaimed
Keep same as SaleVesting.

File: contracts/Vault.sol [hunk 4]
```diff
@@ -1,5 +1,4 @@
-pragma solidity ^0.8.11;
-pragma abicoder v2;
+pragma solidity ^0.8.4;
```
Comment (xiangzhe0812) [inline contracts/Vault.sol:1]:
```suggestion
pragma solidity 0.8.4;
```

File: contracts/SaleVesting.sol [hunk 5]
```diff
@@ -32,11 +31,22 @@ contract SaleVesting is Ownable {
...
     constructor(address _RITE, uint256 _TGEDate) {
+        TGEDate = _TGEDate;
+        deployDate = block.timestamp;
+        require(
+            TGEDate >= deployDate,
+            "TGE cannot be before the deployment date"
+        );
+        require(_RITE != address(0), "Address cannot be zero");
         self = address(this);
         RITE = _RITE;
```
[the remark this answers is not in the document]
Comment (jrocca82) [reply contracts/SaleVesting.sol:49]: This order causes tests to return an empty string, I think I need to have the variables before the require statements

File: contracts/TeamVesting.sol (hunk 2 above)
Comment (jrocca82) [reply contracts/TeamVesting.sol:48]: Need variables first or tests return an empty string

File: contracts/SaleVesting.sol [hunk 6]
```diff
@@ -34,15 +33,24 @@ contract SaleVesting is Ownable {
     uint256 public TGEDate;
     uint256 public totalClaimed;
     uint256 public totalVestingAmount = 0;
+    //Maximun number of vesting detail array.
+    uint256 internal immutable maxVestingDetailArray = 20;
```
Comment (bitcoinbrisbane) [inline contracts/SaleVesting.sol:37]: If we're not setting in the contructor, lets make this constant.  Also, lets change the scope to private if this contract is not being inherited.  Linting might complain its not uppercase...

```suggestion
    uint256 private const maxVestingDetailArray = 20;
```

File: contracts/SaleVesting.sol [hunk 7]
```diff
@@ -64,6 +72,16 @@ contract SaleVesting is Ownable {
         );
 
         for (uint256 i = 0; i < count; i++) {
+            //Check on the maximun size over which the for loop will run over.
+            require(i <= maxVestingDetailArray, "Too many vesting details");
+            //Check there are tokens available
```
Comment (bitcoinbrisbane) [inline contracts/SaleVesting.sol:77]: I see what you're doing after the audit report.  But this will be gas expensive.  Another approach could be:

```suggestion
//Check on the maximun size over which the for loop will run over.
        require(count <= maxVestingDetailArray, "Too many vesting details");

        for (uint256 i = 0; i < count; i++) {
            //Check there are tokens available
```

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Few things to consider

**Review by xiangzhe0812 — APPROVED** [review]
